R/draws-index.R

```diff
@@ -461,14 +463,38 @@ check_existing_variables <- function(variables, x, regex = FALSE,
     missing_variables <- NULL
     variables <- as.character(all_variables[unique(unlist(tmp))])
   } else if (!scalar_only) {
-    missing_candidates <- setdiff(variables, all_variables)
+    # need to find variables that are matched by either a scalar or vector
+    # variable in x and what the matching variable is, while keeping original
+    # order of `variables` => we'll do two left joins to fill in scalar and
+    # vector variable names, which will also find missing variables (as NAs)
+
+    # because the mapping of scalar variables to variables(x) is one-to-one, can
+    # do the first left join using match instead of merge, which is faster
+    scalar_variables <- all_variables[match(variables, all_variables)]
+
+    # need to use merge for the second join as vector variables may match
+    # multiple variables in `x`
     all_variables_base <- gsub("\\[.*\\]$", "", all_variables)
-    missing_variables <- setdiff(missing_candidates, all_variables_base)
-    vector_variables <- setdiff(missing_candidates, missing_variables)
-    variables <- unique(
-      c(variables[!(variables %in% missing_candidates)],
-        all_variables[all_variables_base %in% vector_variables])
+    variables_df <- merge(
+      # left join seems to rearrange input order, so need to keep it
+      # and the order of variables in x so we can restore them later
+      data.frame(
+        variable = variables, input_order = seq_along(variables), scalar = scalar_variables,
+        stringsAsFactors = FALSE
+      ),
```

Comment:
```suggestion
      tibble(
        variable = variables, input_order = seq_along(variables), scalar = scalar_variables
      ),
```

Since *tibble* is imported, it's possible to use it for more reliable default behaviour.

Reply:
Does using a tibble affect performance?

Reply:
I just did some profiling and benchmarking and saw no difference in performance for tibble versus data.frame here (I have seen other cases where tibble slowed stuff down, but it seems not here). I have no strong preference in tibble versus data.frame here so it's up to you.

I did notice a slight improvement using perl regexes in the gsub call there, so I changed that to `perl = TRUE`.

Reply:
I guess `tibble()` construction is a bit slower because of quosures processing etc, but here's it's just 3 columns, so the impact should be minor, unless this code is used in some heavy loops.

Reply:
Out of curiosity I've made a merge-free version :)
I'll publish the branch in ~1h. In theory, it should be faster than the current one.

Reply:
My version is here: https://github.com/alyst/posterior/tree/fix-188.
It's using 2 *match()* calls instead of *match()* and *merge()*, so potentially could be faster.
I haven't run any benchmarks, though.
Also it's possible that replacing *match()* with *vctrs::vec_match()* may speed it up, but I didn't test.